## Implementing Multiple Linear Regression Using Gradient Descent And Stochastic Gradient Descent Algorithm.

In this assignment, you will be implementing Multiple Linear Regression Model on the same data set as in programming assignment 1, but here we will be using gradient descent algorithm (GDA) and stochastic gradient descent algorithm (SGDA) to minimize the cost function as taught in the class. The SGDA implementation is optional.

Please add your own print statements to check your code to ensure your code is correct in every step.  (Note: we will not be grading the print statements you add to your code.)

In [145]:
# run this code cell using shift+enter before moving further
import numpy as np
import pandas as pd
import matplotlib
from matplotlib import pyplot as plt

# Loading the data set.

In the below code cell, you will load the data using python pandas library as done in the programming_assignment_1b. Use pd.read_csv('File url ', header=None,.... ) with the value of header=None,delim_whitespace=True,names=names,na_values='?' as attributes. The url for the .data file is https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data. This step is same as done in programming_assignment_1b.

In [146]:
# After completing the code in this code cell, run this code cell before moving further.

names =[
    'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 
    'AGE',  'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'PRICE'
]

#Write your code below to save dataframe in the df variable below. 
# In place of None, write the pandas command to read the csv file.
df= pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data', sep='\s+', names=names)


Now, instead of taking just 1 feature i.e 'RM' as we did in Simple Linear Regression, we will use 4 features which are 'LSTAT', 'DIS', 'RM' and 'INDUS'. We choose only these features because they are almost on similar scale. So, there is no need for feature scaling. Fetching the values of 5 columns into a smaller dataframe (say) df1 from df. This step is also same as done in programming assignment 1.There we fetched only 2 columns but here we need 5 columns.We need the values in the 'PRICE', 'LSTAT', 'DIS', 'RM' and 'INDUS' columns.

In [147]:
#  After completing the code in this code cell, run this code cell before moving further. Also, save the values in df2 after dropping empty rows from df1
# Write your code below.
df1=df.filter(["LSTAT","DIS","RM","INDUS"],axis=1)
df2=df.filter(["PRICE","LSTAT","DIS","RM","INDUS"],axis=1)
# Check the shape of df2. It should be (506,5)
# print df1.shape
print df2.shape


(506, 5)


Create a vector y having the values of 'PRICE' column and matrix x having the values of  'LSTAT', 'DIS', 'RM' and 'INDUS' columns. We have already written the code for that. Just run the code cell below.

In [148]:
# Just run this code cell
df3 = df2[['LSTAT', 'DIS', 'RM', 'INDUS']]
x=df3.values
y=np.array(df2['PRICE'])
# Check the shape of x and y vectors.
print x.shape
print y.shape


(506, 4)
(506,)


Reshape y to be rank 2. After checking the shape of x and y in the above code cell, we see that x is already rank 2 and y are rank 1 matrix. Before moving ahead, convert y to be rank 2 matrix. For example, I would use the command y=y.reshape(y.shape[0],1) to reshape.

In [149]:
# After completing the code in this code cell, run this code cell before moving further. 
# Write your code below
y = y.reshape(y.shape[0],1)
print y.shape


(506, 1)


Calculating the value of n i.e. number of training examples. 
Hint: Value of n is equal to the number of rows in either x or y matrix which can be accessed using numpy shape command.

In [150]:
# After completing the code in this code cell, run this code cell before moving further. 
# Write your code below
n=x.shape[0]
# After writing a code, it is a good practice to verify that your code is correct. 
# For example, in this case you can print the value of n and check that it should be equal to 506.
# Occasionally we created test code cell to verify your code. 
# However, you should do it for almost every code you write.
print n

506


# Appending a Column of ones in x: Just run the code cell below
Warning: Just run the below code cell only once or you would be appending column of ones multiple times if you run the cell multiple times. If by mistake you run the code cell more than once, re-initilize the value of x again in the previous code cell.

In [151]:
a=np.ones((y.shape[0],1))
x=np.hstack((a , x))

In [158]:
# Shape of x should be (506,5)
print(x.shape)

(506, 5)


# Cost Function.
Compute the cost: Write the code to compute the cost inside the function. Do not change the function name or function parameters.

In [186]:
def compute_cost(x, y, w, n):
    #Write your code in place of None. Cost can be calculated using a single line of code.
    # Remember w is a vector here.
    cost=np.sum(np.square(np.dot(x,w)-y))/(2*n)


    return cost

Before moving ahead, ensure that the code you have written to compute the cost is correct. Just run the below cell.

In [187]:
w_testcase=np.zeros((5,1))
cost_verify= compute_cost(x, y, w_testcase, n)
print(cost_verify)

296.0734584980237


Your output should be equal to 296.073458498. If it's equal to this, then move ahead. Else, re-check your code and re-verify.

# Gradient Descent
Write the code to perform gradient descent in the function below

In [155]:
def gradient_descent(x , y , learning_rate , w , n , num_iters):
    # In place of None, write the updated value of w in temp 
    for i in range(num_iters):
        # derivative vector is given by : X_train.Transpose *  (( X_train * w_vector)- y ) 
        temp =  w - (learning_rate/n)*(np.dot(np.transpose(x),np.dot(x,w)-y))
        w = temp
       
        
        if(i%100==0):
            # In place of None, call the cost you just coded above
            cost= compute_cost(x, y, w, n)
            print("Cost")
            print(cost)
             
            
            
    return w      

Before moving ahead, ensure that your code to update w is correct. Run the code cell below.

In [156]:
w_testcase=np.zeros((5,1))
g=gradient_descent(x , y , 0.0049 , w_testcase , n , 100000)
print(g[0])

Cost
267.2394686487175
Cost
17.992016338677416
Cost
15.29348823112128
Cost
14.678638998273042
Cost
14.52724934213347
Cost
14.48962515076753
Cost
14.479943796087383
Cost
14.477127192952596
Cost
14.475998525491512
Cost
14.47528635184695
Cost
14.474678395546798
Cost
14.474097952523573
Cost
14.473526175661238
Cost
14.472958427629937
Cost
14.472393561080688
Cost
14.471831285845981
Cost
14.471271522320357
Cost
14.470714242649631
Cost
14.470159431726556
Cost
14.469607077611158
Cost
14.469057169177788
Cost
14.468509695537064
Cost
14.467964645893748
Cost
14.467422009511692
Cost
14.466881775705046
Cost
14.46634393383595
Cost
14.46580847331381
Cost
14.465275383594975
Cost
14.464744654182478
Cost
14.464216274625834
Cost
14.463690234520831
Cost
14.463166523509317
Cost
14.462645131279007
Cost
14.462126047563274
Cost
14.46160926214094
Cost
14.461094764836094
Cost
14.460582545517873
Cost
14.460072594100271
Cost
14.459564900541935
Cost
14.459059454845981
Cost
14.458556247059777
Cost
14.458055267274762


The last output should be: 7.46236374. If it matches your out, then your code is likely to be correct. Otherwise, re-check your code and re-verify.

# Integrating the Batch Gradient Descent Function 

Integrating the above function into a single function multiple_linear_reg_model_gda: This function uses gradient descent algorithm to minimize the cost.

In [190]:
def multiple_linear_reg_model_gda(x , y , n , learning_rate , num_iters):
    #initialize the values of parameter vector w. It should be a column vector of zeros of dimension(n,1)
    w =np.zeros((n,1),dtype=int)
    
    #calculate the initial cost by calling the function you coded above.
    xt=np.transpose(x)
    print xt
    initial_cost= compute_cost(xt, y, w, n)
    print("Initial Cost")
    print(initial_cost)
    
    #calculate the optimized value of gradients by calling the gradient_descent function coded above
    
    w = gradient_descent(x , y , learning_rate , w , n , num_iters)
    
    #Calculate the cost with the optimized value of w by calling the cost function.
    
    final_cost = compute_cost(x, y, w, n)
    print("Final Cost")
    print(final_cost)
    return w

Now, when you have coded multiple_linear_reg_model_gda function, you can call this function to find the optimized values of parameters w. Before calling the function, set the values of learning_rate and num_iters. You may have to call this function several number of times with different values of num_iters and learning_rate to find the optimal values of w. For some values of learning_rate, it may give an error as the values of cost may reach a very high value(infinity) due to overshooting as discussed in the class. 

In [191]:
# Write your code below
learning_rate = .003
num_iters = 100000
# In place of None, call the multiple_linear_reg_model_gda.
w = multiple_linear_reg_model_gda(x , y , n , learning_rate , num_iters)

# The value of final cost should be 14.3470049896 or nearly this(depending on the values of learning_rate and num_itersations you choose.)

[[ 1.      1.      1.     ...  1.      1.      1.    ]
 [ 4.98    9.14    4.03   ...  5.64    6.48    7.88  ]
 [ 4.09    4.9671  4.9671 ...  2.1675  2.3889  2.505 ]
 [ 6.575   6.421   7.185  ...  6.976   6.794   6.03  ]
 [ 2.31    7.07    7.07   ... 11.93   11.93   11.93  ]]


ValueError: operands could not be broadcast together with shapes (5,1) (506,1) 

# Normal Equation Method
Now, we will be writing the code to find the values of parameters w for our multiple linear regression model. This can also be used to cross-check the optimal values of w we just found above using method above. These values should be same (or nearly same).
Instead of writing the code for normal equation in one line, you can break this into 3 parts: First calculate q=inverse of (dot of (X.T,X)) (these are pseudo commands, use original numpy commands to calculate q). Then w= dot of ( X.T , y) and then beta_vec= dot of (q,w). Here, w_vec is vector of dimension (5,1) having two values. Example w0=w_vec[0][0].

Note : Do not append a column of ones in x because you have already done before. 

In [ ]:
# Write the code below
q = None
w = None
w_vec = None


In [ ]:
print(w)

The Values of beta you just got above should be approximately same as the ones you got using multiple_linear_reg_model_gda.
This assignment ends here.

# Stochastic Gradient Descent (Optional / Ungraded)

In [ ]:
def stochastic_gradient_descent(x , y , learning_rate , w , n , num_iters):

    for j in range(num_iters):
    
        for i in range(0,n):
        # Write updated value of w vector in place of None
            temp = None 
        
            w = temp
   

        if(j%2000==0):
            cost= None
            print("Cost")
            print(cost)
            
            
    return w

In [ ]:
def multiple_linear_reg_model_sgda(x , y , n , learning_rate , num_iters):
    #initialize the values of parameter vector w. It should be a column vector of zeros of dimension(m,1)
    w = None
    
    #calculate the initial cost by calling the function you coded above.
    initial_cost=None
    print("Initial Cost")
    print(initial_cost)
    
    #calculate the optimized value of gradients by calling the stochastic_gradient_descent function coded above
    
    w = None
    
    #Calculate the cost with the optimized value of w by calling the cost function.
    
    final_cost=None
    print("Final Cost")
    print(final_cost)
    return w

Now, when you have coded multiple_linear_reg_model_sgda function, you can call this function to find the optimized values of parameters w. Before calling the function, set the values of learning_rate and num_iters. You may have to call this function several number of times with different values of num_iters and learning_rate to find the optimal values of parameters w. For some values of learning_rate, it may give an error as the values of parameters may reach a very high value(infinity).

In [ ]:
# Write your code below
learning_rate=None
num_iters=None
# In place of None, call the function multiple_linear_reg_model_sgda.
w=None